### PR TITLE
Allow Airflow UI to create worker pod via Clear > Run

### PR DESCRIPTION
**Pull request by kimyen**
Problem:

Clear and Run task instance via UI cause mushroom cloud error.

Repro steps:
- Using Helm 1.1.0 to deploy Airflow 1.10.15
- From the UI, navigate to a DAG tree view
- Select an existing task (small square on tree view) > Clear > Run

Error:
```
-------------------------------------------------------------------------------
Node: airflow-webserver-67bf9669bc-x6wcc
-------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  ...
  File "/usr/local/lib/python3.7/site-packages/airflow/www/views.py", line 1196, in run
    executor.start()
  File "/usr/local/lib/python3.7/site-packages/airflow/executors/kubernetes_executor.py", line 784, in start
    self.kube_client, self.worker_uuid
  File "/usr/local/lib/python3.7/site-packages/airflow/executors/kubernetes_executor.py", line 396, in __init__
    self.worker_configuration_pod = WorkerConfiguration(kube_config=self.kube_config).as_pod()
  File "/usr/local/lib/python3.7/site-packages/airflow/kubernetes/worker_configuration.py", line 437, in as_pod
    return PodGenerator(pod_template_file=self.kube_config.pod_template_file).gen_pod()
  File "/usr/local/lib/python3.7/site-packages/airflow/kubernetes/pod_generator.py", line 308, in gen_pod
    result.metadata.name = self.make_unique_pod_id(result.metadata.name)
AttributeError: 'NoneType' object has no attribute 'name'
```

How does this PR fix the problem above:

- From the stack trace above, this line here shows that the name is read from the kube_config.pod_template_file: https://github.com/apache/airflow/blob/db720743c5f5828fabed9dd88384c86a41e2f997/airflow/kubernetes/worker_configuration.py#L437 
- In the Airflow Helm chart 1.1.0, only the scheduler has the pod_template_file mounted. So when this action was triggered, the webserver does not have the pod_template file mounted and couldn't fill in the values here. 
- This PR patches the webserver K8s template to also mount the pod_template_file the same way the schedule does here: https://github.com/apache/airflow/blob/daa725b9f359c23a374a808ad53bbd7ecefce0d9/chart/templates/scheduler/scheduler-deployment.yaml#L141 so that when the code path is exercised, `name` is filled in from the value read in the pod template file.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: chart/templates/webserver/webserver-deployment.yaml [hunk 1]
```diff
@@ -129,6 +129,12 @@ spec:
           resources:
 {{ toYaml .Values.webserver.resources | indent 12 }}
           volumeMounts:
+{{- if semverCompare ">=1.10.12" .Values.airflowVersion }}
```
Comment (potiuk) [inline chart/templates/webserver/webserver-deployment.yaml:135]: Should it be also <2.0.0 ?

Comment (potiuk) [reply chart/templates/webserver/webserver-deployment.yaml:135]: Also @jedcunningham @kaxil  -> I am not sure what will be the policy about supporting 1.10 in the chart (See https://apache-airflow.slack.com/archives/CCV3FV9KL/p1631564642172400 for context of that change)

Comment (jedcunningham) [reply chart/templates/webserver/webserver-deployment.yaml:135]: It probably should be present for 2.0.0+ too, as it is config, even if it isn't being used.

**Comment by kimyen** [comment]
One of the unit test failed:
```
==== Helm sqlite: 2 failures ====

chart/tests/test_webserver.py::WebserverDeploymentTest::test_logs_persistence_adds_volume_and_mount_1: AssertionError: assert {'mountPath':...name': 'logs'} == {'mountPath':...'airflow.cfg'}
  Differing items:
  {'name': 'logs'} != {'name': 'config'}
  {'mountPath': '/opt/airflow/logs'} != {'mountPath': '/opt/airflow/airflow.cfg'}
  Right contains 2 more items:
  {'readOnly': True, 'subPath': 'airflow.cfg'}
  Use -v to get the full diff
chart/tests/test_webserver.py::WebserverDeploymentTest::test_logs_persistence_adds_volume_and_mount_2: AssertionError: assert {'mountPath':...name': 'logs'} == {'mountPath':...'airflow.cfg'}
  Differing items:
  {'name': 'logs'} != {'name': 'config'}
  {'mountPath': '/opt/airflow/logs'} != {'mountPath': '/opt/airflow/airflow.cfg'}
  Right contains 2 more items:
  {'readOnly': True, 'subPath': 'airflow.cfg'}
  Use -v to get the full diff
Error: Process completed with exit code 1.
```

I want to challenge the value of this test. It doesn't seem to catch anything that would break the `webserver` deployment, but rather being picky about the order of the volume. 😞 

I would expect to break a test that has a set of volumes that the webserver deployment should have, and with this PR, extend the test to add the new volume. I don't know enough to write such a test. What do you think about the proposal? Does such test already exist (if so, it didn't fail)? Can I get some pointers on how to write the test I propose?

File: chart/templates/webserver/webserver-deployment.yaml (hunk 1 above)
Comment (kaxil) [reply chart/templates/webserver/webserver-deployment.yaml:135]: Currently, there is not much maintenance effort for supporting 1.x so I think we should support it for now

**Comment by jedcunningham** [comment]
@kimyen, that test is overly fragile as written. Easy fix though: #18332

I don't think that test exists yet, and I agree it would be a valuable test. The broken test is a reasonable example to follow, just check the default `volumes` and `volumeMounts` instead. More than 1 way to approach this, but one would be to use parameterized.expand like so:
```
@parameterized.expand([
    ("1.10.10", False),
    ("1.10.12", True),
    ("2.1.0", True),
])
def test_default_volumes_and_mounts(self, af_version, pod_template_file_expected):
    # assert "always there" volumes/mounts
    if pod_template_file_expected:
        # assert pod_template_file mount   
```

**Comment by kimyen** [comment]
@jedcunningham I would love your feedback on the test that I added. Other volumes depends on other values, so I limit the scope of my test on default and optional config volumes. 

I don't have an option to add tags to this PR, but would like to mark it as `Ready for Review`.

File: chart/tests/test_webserver.py [hunk 2]
```diff
@@ -236,6 +236,36 @@ def test_logs_persistence_adds_volume_and_mount(self, log_persistence_values, ex
...
+        } in jmespath.search("spec.template.spec.containers[0].volumeMounts", docs[0])
+
+        # pod_template_file config
+        assert (
+            pod_template_file_expected
+            == {
+                "name": "config",
+                "mountPath": "/opt/airflow/pod_template_file.yaml",
+            }
+            in jmespath.search("spec.template.spec.containers[0].volumeMounts", docs[0])
```
Comment (jedcunningham) [inline chart/tests/test_webserver.py]:
```suggestion
            == (
                {
                    "name": "config",
                    "mountPath": "/opt/airflow/pod_template_file.yaml",
                }
                in jmespath.search("spec.template.spec.containers[0].volumeMounts", docs[0])
            )
```

Not sure on the formatting here, but I believe we need parens.

File: chart/tests/test_webserver.py [hunk 3]
```diff
@@ -236,6 +236,38 @@ def test_logs_persistence_adds_volume_and_mount(self, log_persistence_values, ex
...
+        assert (
+            pod_template_file_expected
+            == (
+                {
+                    "name": "config",
+                    "mountPath": "/opt/airflow/pod_template_file.yaml",
+                }
+                in jmespath.search("spec.template.spec.containers[0].volumeMounts", docs[0])
+            )
+        )
```
Comment (kimyen) [inline chart/tests/test_webserver.py]: Use CI suggested format:
```suggestion
        assert pod_template_file_expected == (
                {
                    "name": "config",
                    "mountPath": "/opt/airflow/pod_template_file.yaml",
                }
                in jmespath.search("spec.template.spec.containers[0].volumeMounts", docs[0])
        )
```

File: chart/tests/test_webserver.py [hunk 4]
```diff
@@ -236,6 +236,35 @@ def test_logs_persistence_adds_volume_and_mount(self, log_persistence_values, ex
...
+            "mountPath": "/opt/airflow/airflow.cfg",
+        } in jmespath.search("spec.template.spec.containers[0].volumeMounts", docs[0])
+
+        # pod_template_file config
+        assert pod_template_file_expected == (
+                {
+                    "name": "config",
+                    "mountPath": "/opt/airflow/pod_template_file.yaml",
+                }
+                in jmespath.search("spec.template.spec.containers[0].volumeMounts", docs[0])
```
Comment (jedcunningham) [inline chart/tests/test_webserver.py]:
```suggestion
            {
                "name": "config",
                "mountPath": "/opt/airflow/pod_template_file.yaml",
            }
            in jmespath.search("spec.template.spec.containers[0].volumeMounts", docs[0])
```

**Comment by jedcunningham** [comment]
@kimyen I'd also suggest that you [set up the pre-commit hooks](https://github.com/apache/airflow/blob/main/STATIC_CODE_CHECKS.rst#prerequisites-for-pre-commit-hooks), they help catch these static check issues locally 👍.

**Comment by kimyen** [comment]
Screen Shot 2021-10-04 at 7 58 09 AM

**Comment by potiuk** [comment]
REopened to rebuild @kimyen -> Next time you can simply rebase + "--force-push-with-lease" if something like that happens  (which might be because of various CI issues)

**Comment by kimyen** [comment]
Thanks @potiuk !

The docker image build still failed. But the error is not related to this change:
```
  ERROR: Cannot install apache-airflow[devel-ci]==2.2.0.dev0 because these package versions have conflicting dependencies.
```

It looks like the fix is in package dependencies for 2.2.0.dev0.

**Comment by potiuk** [comment]
Looks like all tests passed now!

**Review by jedcunningham — APPROVED** [review]

**Pull request merged**